File: poly_apps/nuxt_main/apps/app_codemart/types/codemart-enums.ts
/**
 * CodeMart Centralized Enums and Constants
 * Single source of truth for all data types and enumerations
 * Used across all APIs to prevent data duplication
 */

// ============ USER ROLES ============
export enum UserRoleType {
  DEVELOPER = 'developer',
  CLIENT = 'client',
  ARCHITECT = 'architect',
  REVIEWER = 'reviewer',
  ADMIN = 'admin',
}

export enum UserRoleStatus {
  PENDING = 'pending',
  APPROVED = 'approved',
  ACTIVE = 'active',
  SUSPENDED = 'suspended',
  REJECTED = 'rejected',
}

// ============ VERIFICATION STATUS ============
export enum VerificationStatus {
  NOT_STARTED = 'not_started',
  PENDING = 'pending',
  APPROVED = 'approved',
  REJECTED = 'rejected',
}

export enum EmailVerificationStatus {
  UNVERIFIED = 'unverified',
  VERIFIED = 'verified',
  EXPIRED = 'expired',
}

export enum PhoneVerificationStatus {
  UNVERIFIED = 'unverified',
  VERIFIED = 'verified',
}

export enum IdentityType {
  ID_CARD = 'ID_CARD',
  PASSPORT = 'PASSPORT',
  DRIVING_LICENSE = 'DRIVING_LICENSE',
}

// ============ PROJECT MANAGEMENT ============
export enum ProjectStatus {
  DRAFT = 'draft',
  OPEN = 'open',
  IN_PROGRESS = 'in_progress',
  PAUSED = 'paused',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
  ARCHIVED = 'archived',
}

export enum ProjectBudgetRange {
  UNDER_5K = 'under_5k',
  RANGE_5K_10K = '5k_10k',
  RANGE_10K_50K = '10k_50k',
  RANGE_50K_100K = '50k_100k',
  OVER_100K = 'over_100k',
}

export enum ProjectComplexity {
  SIMPLE = 'simple',
  MEDIUM = 'medium',
  COMPLEX = 'complex',
  VERY_COMPLEX = 'very_complex',
}

export enum MilestoneStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

// ============ TASK MANAGEMENT ============
export enum TaskStatus {
  OPEN = 'open',
  ASSIGNED = 'assigned',
  IN_PROGRESS = 'in_progress',
  SUBMITTED = 'submitted',
  UNDER_REVIEW = 'under_review',
  APPROVED = 'approved',
  REJECTED = 'rejected',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
}

export enum TaskPriority {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  URGENT = 'urgent',
}

export enum TaskCommentType {
  GENERAL = 'general',
  CODE_REVIEW = 'code_review',
  QUESTION = 'question',
  SUGGESTION = 'suggestion',
}

// ============ PAYMENT & WALLET ============
export enum PaymentMethod {
  ALIPAY = 'alipay',
  WECHAT = 'wechat',
  BANK_TRANSFER = 'bank_transfer',
  CREDIT_CARD = 'credit_card',
  WALLET = 'wallet',
}

export enum PaymentStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  REFUNDED = 'refunded',
}

export enum PaymentType {
  DEPOSIT = 'deposit',
  PROJECT_PAYMENT = 'project_payment',
  MILESTONE_PAYMENT = 'milestone_payment',
  REFUND = 'refund',
  WITHDRAWAL = 'withdrawal',
  BONUS = 'bonus',
}

export enum TransactionStatus {
  PENDING = 'pending',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

export enum WalletTransactionType {
  DEPOSIT = 'deposit',
  WITHDRAWAL = 'withdrawal',
  PROJECT_EARNING = 'project_earning',
  BONUS = 'bonus',
  PENALTY = 'penalty',
  REFUND = 'refund',
}

// ============ CODE REVIEW & RATING ============
export enum CodeQualityScore {
  POOR = 1,
  FAIR = 2,
  GOOD = 3,
  VERY_GOOD = 4,
  EXCELLENT = 5,
}

export enum ReviewStatus {
  PENDING = 'pending',
  IN_REVIEW = 'in_review',
  COMPLETED = 'completed',
  DISPUTED = 'disputed',
}

// ============ SKILL & CERTIFICATION ============
export enum SkillLevel {
  BEGINNER = 'beginner',
  INTERMEDIATE = 'intermediate',
  ADVANCED = 'advanced',
  EXPERT = 'expert',
}

export enum CertificationStatus {
  PENDING = 'pending',
  VERIFIED = 'verified',
  EXPIRED = 'expired',
  REJECTED = 'rejected',
}

// ============ NOTIFICATION ============
export enum NotificationType {
  PROJECT_INVITATION = 'project_invitation',
  TASK_ASSIGNED = 'task_assigned',
  PAYMENT_RECEIVED = 'payment_received',
  SUBMISSION_APPROVED = 'submission_approved',
  SUBMISSION_REJECTED = 'submission_rejected',
  MESSAGE = 'message',
  SYSTEM = 'system',
}

export enum NotificationStatus {
  UNREAD = 'unread',
  READ = 'read',
  ARCHIVED = 'archived',
}

// ============ DEVELOPER LEVEL & PROMOTION ============
export enum DeveloperLevel {
  JUNIOR = 'junior',
  INTERMEDIATE = 'intermediate',
  SENIOR = 'senior',
  EXPERT = 'expert',
  ARCHITECT = 'architect',
}

// ============ INDUSTRY & TECH STACK ============
export enum Industry {
  SAAS = 'saas',
  ECOMMERCE = 'ecommerce',
  FINTECH = 'fintech',
  HEALTHCARE = 'healthcare',
  EDUCATION = 'education',
  ENTERPRISE = 'enterprise',
  MOBILE = 'mobile',
  IOT = 'iot',
  AI_ML = 'ai_ml',
  OTHER = 'other',
}

export enum ProgrammingLanguage {
  JAVASCRIPT = 'javascript',
  TYPESCRIPT = 'typescript',
  PYTHON = 'python',
  JAVA = 'java',
  PHP = 'php',
  CSHARP = 'csharp',
  GOLANG = 'golang',
  RUST = 'rust',
  RUBY = 'ruby',
  CPP = 'cpp',
  KOTLIN = 'kotlin',
  SWIFT = 'swift',
}

export enum Framework {
  // Frontend
  REACT = 'react',
  VUE = 'vue',
  ANGULAR = 'angular',
  SVELTE = 'svelte',

  // Backend
  LARAVEL = 'laravel',
  DJANGO = 'django',
  EXPRESS = 'express',
  SPRING = 'spring',
  DOTNET = 'dotnet',
  GIN = 'gin',
  FASTAPI = 'fastapi',

  // Full Stack
  NEXTJS = 'nextjs',
  NUXT = 'nuxt',
  REMIX = 'remix',
}

export enum Database {
  MYSQL = 'mysql',
  POSTGRESQL = 'postgresql',
  MONGODB = 'mongodb',
  REDIS = 'redis',
  ELASTICSEARCH = 'elasticsearch',
  DYNAMODB = 'dynamodb',
  FIRESTORE = 'firestore',
  CASSANDRA = 'cassandra',
  MARIADB = 'mariadb',
}

// ============ CONSTANTS ============
export const DEFAULT_DEPOSIT_AMOUNTS = {
  [UserRoleType.DEVELOPER]: 199.00,
  [UserRoleType.CLIENT]: 0.00,
  [UserRoleType.ARCHITECT]: 999.00,
};

export const PAGINATION_DEFAULTS = {
  PAGE: 1,
  PAGE_SIZE: 20,
  MAX_PAGE_SIZE: 100,
};

export const FILE_UPLOAD_LIMITS = {
  MAX_SIZE_MB: 10,
  IMAGE_MAX_SIZE_MB: 5,
  SUPPORTED_IMAGES: ['jpg', 'jpeg', 'png', 'gif', 'webp'],
  SUPPORTED_DOCUMENTS: ['pdf', 'doc', 'docx', 'xls', 'xlsx'],
};

export const OTP_CONFIG = {
  LENGTH: 6,
  EXPIRY_MINUTES: 10,
  MAX_ATTEMPTS: 5,
  RESEND_COOLDOWN_SECONDS: 60,
};

export const EMAIL_CONFIG = {
  TOKEN_LENGTH: 64,
  EXPIRY_HOURS: 24,
};

export const PROJECT_DEFAULTS = {
  MIN_BUDGET: 100,
  MAX_BUDGET: 1000000,
  MIN_DURATION_DAYS: 1,
  MAX_DURATION_DAYS: 365,
};

export const RATING_SCALE = {
  MIN: 1,
  MAX: 5,
  PRECISION: 2,
};

File: poly_apps/flutter_bloom/lib/apps/app_codemart/models_app_codemart/codemart_enums.dart
enum UserRoleType {
  developer,
  client,
  architect,
  reviewer,
  admin,
}

// Developer sub-types
enum DeveloperType {
  regular,      // Regular developer
  architect,    // Architect
}

// Client sub-types
enum ClientType {
  individual,   // Individual
  enterprise,   // Enterprise
}

enum UserRoleStatus {
  pending,
  approved,
  active,
  suspended,
  rejected,
}

enum VerificationStatus {
  notStarted,
  pending,
  approved,
  rejected,
}

enum EmailVerificationStatus {
  unverified,
  verified,
  expired,
}

enum PhoneVerificationStatus {
  unverified,
  verified,
}

enum IdentityType {
  idCard,
  passport,
  drivingLicense,
}

enum ProjectStatus {
  draft,
  open,
  inProgress,
  paused,
  completed,
  cancelled,
  archived,
}

enum ProjectBudgetRange {
  under5k,
  range5k10k,
  range10k50k,
  range50k100k,
  over100k,
}

enum ProjectComplexity {
  simple,
  medium,
  complex,
  veryComplex,
}

enum MilestoneStatus {
  pending,
  inProgress,
  completed,
  failed,
  cancelled,
}

enum TaskStatus {
  open,
  assigned,
  inProgress,
  submitted,
  underReview,
  approved,
  rejected,
  completed,
  cancelled,
}

enum TaskPriority {
  low,
  medium,
  high,
  urgent,
}

enum TaskCommentType {
  general,
  codeReview,
  question,
  suggestion,
}

enum PaymentMethod {
  alipay,
  wechat,
  bankTransfer,
  creditCard,
  wallet,
}

enum PaymentStatus {
  pending,
  processing,
  completed,
  failed,
  cancelled,
  refunded,
}

enum PaymentType {
  deposit,
  projectPayment,
  milestonePayment,
  refund,
  withdrawal,
  bonus,
}

enum TransactionStatus {
  pending,
  completed,
  failed,
  cancelled,
}

enum WalletTransactionType {
  deposit,
  withdrawal,
  projectEarning,
  bonus,
  penalty,
  refund,
}

enum CodeQualityScore {
  poor,
  fair,
  good,
  veryGood,
  excellent,
}

enum ReviewStatus {
  pending,
  inReview,
  completed,
  disputed,
}

enum SkillLevel {
  beginner,
  intermediate,
  advanced,
  expert,
}

enum CertificationStatus {
  pending,
  verified,
  expired,
  rejected,
}

enum NotificationType {
  projectInvitation,
  taskAssigned,
  paymentReceived,
  submissionApproved,
  submissionRejected,
  message,
  system,
}

enum NotificationStatus {
  unread,
  read,
  archived,
}

// Developer levels (based on points and completed projects)
enum DeveloperLevel {
  level1,    // Newbie (0-100 points)
  level2,    // Junior (101-500 points)
  level3,    // Intermediate (501-1500 points)
  level4,    // Senior (1501-3000 points)
  level5,    // Expert (3001-5000 points)
  level6,    // Master (5001+ points)
}

// Client levels (based on projects posted and payment amount)
enum ClientLevel {
  level1,    // New (0-5 projects)
  level2,    // Bronze (6-15 projects)
  level3,    // Silver (16-30 projects)
  level4,    // Gold (31-50 projects)
  level5,    // Platinum (51-100 projects)
  level6,    // Diamond (100+ projects)
}

enum Industry {
  saas,
  ecommerce,
  fintech,
  healthcare,
  education,
  enterprise,
  mobile,
  iot,
  aiMl,
  other,
}

enum ProgrammingLanguage {
  javascript,
  typescript,
  python,
  java,
  php,
  csharp,
  golang,
  rust,
  ruby,
  cpp,
  kotlin,
  swift,
}

enum Framework {
  react,
  vue,
  angular,
  svelte,
  laravel,
  django,
  express,
  spring,
  dotnet,
  gin,
  fastapi,
  nextjs,
  nuxt,
  remix,
}

enum Database {
  mysql,
  postgresql,
  mongodb,
  redis,
  elasticsearch,
  dynamodb,
  firestore,
  cassandra,
  mariadb,
}

// Attachment types
enum AttachmentType {
  projectDocument,
  taskDocument,
  proposalDocument,
  submissionDocument,
  verificationDocument,
  other,
}

// File types
enum AttachmentFileType {
  pdf,
  doc,
  docx,
  xls,
  xlsx,
  ppt,
  pptx,
  txt,
  zip,
  rar,
  image,
}

// Certification types
enum CertificationType {
  education,        // Educational certificate
  professional,     // Professional certification
  skill,           // Skill certification
  project,         // Project completion certificate
  award,           // Award or recognition
}

// Budget types
enum BudgetType {
  fixed,           // Fixed price
  hourly,          // Hourly rate
  milestone,       // Milestone-based
}

// Project complexity types (detailed)
enum ProjectComplexityType {
  simple,
  medium,
  complex,
  veryComplex,
}

// Project status types (detailed)
enum ProjectStatusType {
  draft,
  open,
  inProgress,
  paused,
  completed,
  cancelled,
  archived,
}

class CodeMartConstants {
  static const Map<UserRoleType, double> defaultDepositAmounts = {
    UserRoleType.developer: 199.00,
    UserRoleType.client: 0.00,
    UserRoleType.architect: 999.00,
  };

  static const int defaultPage = 1;
  static const int defaultPageSize = 20;
  static const int maxPageSize = 100;

  static const int maxFileSizeMb = 10;
  static const int imageMaxSizeMb = 5;
  static const List<String> supportedImages = ['jpg', 'jpeg', 'png', 'gif', 'webp'];
  static const List<String> supportedDocuments = ['pdf', 'doc', 'docx', 'xls', 'xlsx'];

  static const int otpLength = 6;
  static const int otpExpiryMinutes = 10;
  static const int otpMaxAttempts = 5;
  static const int otpResendCooldownSeconds = 60;

  static const int emailTokenLength = 64;
  static const int emailExpiryHours = 24;

  static const double minBudget = 100;
  static const double maxBudget = 1000000;
  static const int minDurationDays = 1;
  static const int maxDurationDays = 365;

  static const int ratingMin = 1;
  static const int ratingMax = 5;
  static const int ratingPrecision = 2;
}
